File: go/cli_signer.go
package remitmd

import (
	"bytes"
	"context"
	"encoding/hex"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// CliSigner implements Signer by delegating to the `remit sign` CLI subprocess.
// The CLI holds the encrypted keystore; this adapter only needs the binary on
// PATH and the REMIT_KEY_PASSWORD env var set.
//
// Create with NewCliSigner, which fetches and caches the wallet address:
//
//	signer, err := remitmd.NewCliSigner()
//	wallet, err := remitmd.NewWalletWithSigner(signer)
type CliSigner struct {
	cliPath string
	address common.Address
}

// cliTimeout is the maximum time allowed for a CLI subprocess call.
const cliTimeout = 10 * time.Second

// NewCliSigner creates a CliSigner by running `remit address` to fetch
// and cache the wallet address. The CLI must be on PATH (or specify a
// custom path), the keystore must exist, and REMIT_KEY_PASSWORD must be set.
func NewCliSigner(cliPath ...string) (*CliSigner, error) {
	path := "remit"
	if len(cliPath) > 0 && cliPath[0] != "" {
		path = cliPath[0]
	}

	ctx, cancel := context.WithTimeout(context.Background(), cliTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, path, "address")
	out, err := cmd.Output()
	if err != nil {
		return nil, remitErr(ErrCodeUnauthorized,
			fmt.Sprintf("CliSigner: failed to get address from CLI: %s", err),
			map[string]any{"cli_path": path},
		)
	}

	addr := strings.TrimSpace(string(out))
	if !strings.HasPrefix(addr, "0x") || len(addr) != 42 {
		return nil, remitErr(ErrCodeUnauthorized,
			fmt.Sprintf("CliSigner: invalid address from CLI: %s", addr),
			map[string]any{"cli_path": path},
		)
	}

	return &CliSigner{
		cliPath: path,
		address: common.HexToAddress(addr),
	}, nil
}

// Sign sends the digest to `remit sign --digest` via stdin and returns
// the 65-byte signature.
func (c *CliSigner) Sign(digest [32]byte) ([]byte, error) {
	hexInput := "0x" + hex.EncodeToString(digest[:])

	ctx, cancel := context.WithTimeout(context.Background(), cliTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, c.cliPath, "sign", "--digest")
	cmd.Stdin = bytes.NewReader([]byte(hexInput))

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		errMsg := strings.TrimSpace(stderr.String())
		if errMsg == "" {
			errMsg = err.Error()
		}
		return nil, remitErr(ErrCodeUnauthorized,
			fmt.Sprintf("CliSigner: signing failed: %s", errMsg),
			nil,
		)
	}

	sigHex := strings.TrimSpace(stdout.String())
	if !strings.HasPrefix(sigHex, "0x") || len(sigHex) != 132 {
		return nil, remitErr(ErrCodeServerError,
			fmt.Sprintf("CliSigner: invalid signature from CLI: %s", sigHex),
			nil,
		)
	}

	sigBytes, err := hex.DecodeString(sigHex[2:])
	if err != nil {
		return nil, remitErr(ErrCodeServerError,
			fmt.Sprintf("CliSigner: invalid signature hex: %s", err),
			nil,
		)
	}
	if len(sigBytes) != 65 {
		return nil, remitErr(ErrCodeServerError,
			fmt.Sprintf("CliSigner: expected 65-byte signature, got %d bytes", len(sigBytes)),
			map[string]any{"length": len(sigBytes)},
		)
	}

	return sigBytes, nil
}

// Address returns the cached Ethereum address of the signing key.
func (c *CliSigner) Address() common.Address {
	return c.address
}

// String returns a safe representation.
func (c *CliSigner) String() string {
	return fmt.Sprintf("CliSigner{address: %s}", c.address.Hex())
}

// IsCliSignerAvailable checks all three conditions for CliSigner activation:
//  1. `remit` (or `remit.exe`) found on PATH
//  2. Keystore file exists at ~/.remit/keys/default.enc
//  3. REMIT_KEY_PASSWORD env var is set (non-empty)
func IsCliSignerAvailable(cliPath ...string) bool {
	path := "remit"
	if len(cliPath) > 0 && cliPath[0] != "" {
		path = cliPath[0]
	}

	// Check 1: CLI on PATH
	if _, err := exec.LookPath(path); err != nil {
		return false
	}

	// Check 2: Keystore exists
	home, err := os.UserHomeDir()
	if err != nil {
		return false
	}
	keystorePath := filepath.Join(home, ".remit", "keys", "default.enc")
	if _, err := os.Stat(keystorePath); os.IsNotExist(err) {
		return false
	}

	// Check 3: Password available
	if os.Getenv("REMIT_KEY_PASSWORD") == "" {
		return false
	}

	return true
}

// cliInstallHint returns platform-specific install instructions for the CLI.
func cliInstallHint() string {
	switch runtime.GOOS {
	case "darwin":
		return "brew install remit-md/tap/remit"
	case "windows":
		return "winget install remit-md.remit"
	default:
		return "curl -fsSL https://remit.md/install.sh | sh"
	}
}

File: rust/src/cli_signer.rs
//! CLI signer adapter for the `remit sign` subprocess.
//!
//! Delegates digest signing to the Remit CLI binary. The CLI holds the encrypted
//! keystore; this adapter only needs the binary on PATH and the `REMIT_KEY_PASSWORD`
//! env var set.
//!
//! # Usage
//!
//! ```rust,ignore
//! use remitmd::CliSigner;
//! use remitmd::Wallet;
//!
//! let signer = CliSigner::new()?;
//! let wallet = Wallet::with_signer(signer).testnet().build()?;
//! ```

use std::io::Write;
use std::process::{Command, Stdio};

use crate::error::{codes, remit_err, RemitError};
use crate::signer::Signer;

/// Signer backed by the `remit sign` CLI command.
///
/// - No key material in this process — signing happens in a subprocess.
/// - Address is cached at construction time via `remit address`.
/// - `sign()` pipes a hex digest to `remit sign --digest` on stdin.
/// - All errors are explicit — no silent fallbacks.
pub struct CliSigner {
    cli_path: String,
    address: String,
}

impl CliSigner {
    /// Create a `CliSigner`, fetching and caching the wallet address.
    ///
    /// Runs `remit address` to get the address from the keystore (no password needed).
    /// Fails loud if the CLI is not found or the keystore is missing.
    pub fn new() -> Result<Self, RemitError> {
        Self::with_path("remit")
    }

    /// Create a `CliSigner` with a custom CLI binary path.
    pub fn with_path(cli_path: &str) -> Result<Self, RemitError> {
        let output = Command::new(cli_path)
            .arg("address")
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output()
            .map_err(|e| {
                remit_err(
                    codes::UNAUTHORIZED,
                    format!("CliSigner: failed to run `{cli_path} address`: {e}"),
                )
            })?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            return Err(remit_err(
                codes::UNAUTHORIZED,
                format!(
                    "CliSigner: `{cli_path} address` failed: {}",
                    stderr.trim()
                ),
            ));
        }

        let addr = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if !addr.starts_with("0x") || addr.len() != 42 {
            return Err(remit_err(
                codes::UNAUTHORIZED,
                format!("CliSigner: invalid address from CLI: {addr}"),
            ));
        }

        Ok(Self {
            cli_path: cli_path.to_string(),
            address: addr,
        })
    }

    /// Check all three conditions for CliSigner activation:
    /// 1. `remit` found on PATH
    /// 2. Keystore file exists at `~/.remit/keys/default.enc`
    /// 3. `REMIT_KEY_PASSWORD` env var is set (non-empty)
    pub fn is_available() -> bool {
        Self::is_available_with_path("remit")
    }

    /// Check availability with a custom CLI path.
    pub fn is_available_with_path(cli_path: &str) -> bool {
        // Check 1: CLI on PATH
        if which::which(cli_path).is_err() {
            return false;
        }

        // Check 2: Keystore exists
        let keystore_path = match dirs::home_dir() {
            Some(home) => home.join(".remit").join("keys").join("default.enc"),
            None => return false,
        };
        if !keystore_path.exists() {
            return false;
        }

        // Check 3: Password available
        match std::env::var("REMIT_KEY_PASSWORD") {
            Ok(val) if !val.is_empty() => true,
            _ => false,
        }
    }
}

impl Signer for CliSigner {
    fn sign(&self, digest: &[u8; 32]) -> Result<Vec<u8>, RemitError> {
        let hex_input = format!("0x{}", hex::encode(digest));

        let mut child = Command::new(&self.cli_path)
            .args(["sign", "--digest"])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| {
                remit_err(
                    codes::UNAUTHORIZED,
                    format!("CliSigner: failed to spawn `{} sign`: {e}", self.cli_path),
                )
            })?;

        // Write to stdin and close it
        if let Some(mut stdin) = child.stdin.take() {
            stdin.write_all(hex_input.as_bytes()).map_err(|e| {
                remit_err(
                    codes::SERVER_ERROR,
                    format!("CliSigner: failed to write to stdin: {e}"),
                )
            })?;
            // stdin is dropped here, closing the pipe
        }

        let output = child.wait_with_output().map_err(|e| {
            remit_err(
                codes::SERVER_ERROR,
                format!("CliSigner: failed to read output: {e}"),
            )
        })?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            return Err(remit_err(
                codes::UNAUTHORIZED,
                format!("CliSigner: signing failed: {}", stderr.trim()),
            ));
        }

        let sig_hex = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if !sig_hex.starts_with("0x") || sig_hex.len() != 132 {
            return Err(remit_err(
                codes::SERVER_ERROR,
                format!("CliSigner: invalid signature from CLI: {sig_hex}"),
            ));
        }

        let sig_bytes = hex::decode(&sig_hex[2..]).map_err(|e| {
            remit_err(
                codes::SERVER_ERROR,
                format!("CliSigner: invalid signature hex: {e}"),
            )
        })?;

        if sig_bytes.len() != 65 {
            return Err(remit_err(
                codes::SERVER_ERROR,
                format!(
                    "CliSigner: expected 65-byte signature, got {} bytes",
                    sig_bytes.len()
                ),
            ));
        }

        Ok(sig_bytes)
    }

    fn address(&self) -> &str {
        &self.address
    }
}

impl std::fmt::Debug for CliSigner {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CliSigner")
            .field("address", &self.address)
            .finish_non_exhaustive()
    }
}

/// Platform-specific install hint for the CLI.
pub fn cli_install_hint() -> &'static str {
    if cfg!(target_os = "macos") {
        "brew install remit-md/tap/remit"
    } else if cfg!(target_os = "windows") {
        "winget install remit-md.remit"
    } else {
        "curl -fsSL https://remit.md/install.sh | sh"
    }
}
